cpp/src/io/parquet/experimental/page_index_filter.cu

```diff
@@ -668,6 +488,195 @@ struct page_stats_caster : public stats_caster_base {
+ * value. If found, return. Else, move the boundary to its alignment and repeat.
+ *
+ * @param tree_level_ptrs Pointers to the start of Fenwick tree level data
+ * @param page_offsets Pointer to page offsets describing each search range i as [page_offsets[i],
+ * page_offsets[i+1))
+ * @param num_ranges Number of search ranges
+ */
+struct search_fenwick_tree_functor {
+  bool** tree_level_ptrs;
+  cudf::size_type const* page_offsets;
+  cudf::size_type num_ranges;
+
+  /**
+   * @brief Enum class to represent which range boundary to align
+   */
+  enum class alignment : uint8_t {
+    START = 0,
+    END   = 1,
+  };
+
+  /**
+   * @brief Checks if a value is a power of two
+   *
+   * @param value Value to check
+   * @return Boolean indicating if the value is a power of two
+   */
+  __device__ bool inline constexpr is_power_of_two(cudf::size_type value) const noexcept
+  {
+    return (value & (value - 1)) == 0;
+  }
+
+  /**
+   * @brief Finds the smallest power of two in the range [start, end); 0 otherwise
+   *
+   * @param start Range start
+   * @param end Range end
+   * @return Largest power of two in the range [start, end); 0 otherwise
+   */
+  __device__ cudf::size_type inline constexpr smallest_power_of_two_in_range(
+    cudf::size_type start, cudf::size_type end) const noexcept
+  {
+    start--;
+    start |= start >> 1;
+    start |= start >> 2;
+    start |= start >> 4;
+    start |= start >> 8;
+    start |= start >> 16;
+    auto const result = start + 1;
+    return result < end ? result : 0;
+  }
+
+  /**
+   * @brief Finds the largest power of two in the range (start, end]; 0 otherwise
+   *
+   * @param start Range start
+   * @param end Range end
+   * @return Largest power of two in the range (start, end]; 0 otherwise
+   */
+  __device__ size_type inline constexpr largest_power_of_two_in_range(size_type start,
+                                                                      size_type end) const noexcept
+  {
+    auto constexpr nbits = cudf::detail::size_in_bits<size_type>() - 1;
+    auto const result    = size_type{1} << (nbits - cuda::std::countl_zero<uint32_t>(end));
+    return result > start ? result : 0;
+  }
+
+  /**
+   * @brief Aligns a range boundary to the next power-of-two block
+   *
+   * @tparam Alignment The boundary (start or end) to align
+   * @param start Range start
+   * @param end Range end
+   * @return A pair of the tree level and block size
+   */
+  template <alignment Alignment>
+  __device__
+    cuda::std::pair<cudf::size_type, cudf::size_type> inline constexpr align_range_boundary(
+      cudf::size_type start, cudf::size_type end) const noexcept
+  {
+    if constexpr (Alignment == alignment::START) {
+      if (start == 0 or is_power_of_two(start)) {
+        auto const block_size = largest_power_of_two_in_range(start, end);
+        auto const tree_level = cuda::std::countr_zero<uint32_t>(block_size);
+        return cuda::std::pair{tree_level, block_size};
```

Comment:
would you consider returning without the explicit type or using auto for the function return type? I personally don't like to repeat the type as it's more letters and allows accidental implicit conversions if the types don't actually match.